### ISSUE TITLE
Enhancing paths and selectors

### ISSUE THREAD

**Issue by AliYusuf95** [nartc/mapper#299]
Hi @nartc 

My suggestion is to pass a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) in the selector method instead of parsing the method to find selectors, which produced some [limitations](https://automapperts.netlify.app/docs/misc/limitations/) in writing the selector method.

For example by wrapping the destination object with a proxy that handles accessing nested objects and storing the path of it. Eg. by utilizing [proxy-deep](https://www.npmjs.com/package/proxy-deep)
```ts
const source = new DeepProxy({}, {
  get(target, path, receiver) {
    return this.nest(function () {})
  },
  apply(target, thisArg, argList) {
    return this.path;
  }
});

const selector = (destination) =>  destination[`${'fo.o'}`]['fo' + 'o'].bar;

const paths = selector(source)() // result ["fo.o", "foo", "bar"]
```

Also, I think having the path stored as an array is more appropriate than string separated by dots. Current implimentation can't handel this currectly: 

```ts
export class CustomKeyBar {
  @AutoMap({ typeFn: () => String })
  ['.startDot']!: string | null;
}

export class CustomKeyBarVm {
  @AutoMap({ typeFn: () => String })
  ['.startDot']!: string | null;
}

export class CustomKeyFoo {
  @AutoMap()
  ['.startDot']!: string;
  @AutoMap({ typeFn: () => CustomKeyBar })
  ['mid.Dot']!: CustomKeyBar;
  @AutoMap({ typeFn: () => Number })
  ['endDot.']!: number | null;
  @AutoMap({ typeFn: () => String })
  normalKey!: string
}

export class CustomKeyFooVm {
  @AutoMap()
  ['.startDot']!: string;
  @AutoMap({ typeFn: () => CustomKeyBarVm })
  ['mid.Dot']!: CustomKeyBarVm;
  @AutoMap({ typeFn: () => Number })
  ['endDot.']!: number | null;
  @AutoMap({ typeFn: () => String })
  normalKey!: string
}

  it('should map custom keys', () => {
    mapper.createMap(CustomKeyBar, CustomKeyBarVm);
    mapper.createMap(CustomKeyFoo, CustomKeyFooVm);
    const foo = new CustomKeyFoo();
    foo.normalKey = 'Normal';
    foo['.startDot'] = 'Foo';
    foo['endDot.'] = 123;
    foo['mid.Dot'] = new CustomKeyBar();
    foo['mid.Dot']['.startDot'] = 'Bar';

    const vm = mapper.map(foo, CustomKeyFooVm, CustomKeyFoo);
    console.log(vm, foo);
    expect(vm.normalKey).toEqual(foo.normalKey);
    expect(vm['.startDot']).toEqual(foo['.startDot']);
    expect(vm['mid.Dot']['.startDot']).toEqual(foo['mid.Dot']['.startDot']);
    expect(vm['endDot.']).toEqual(foo['endDot.']);
  });
```

**Comment by micalevisk** [comment]
Hi! proxies are cool. I'm already aware of the usage of their with typescript because of [`@golevelup/ts-jest`](https://www.npmjs.com/package/@golevelup/ts-jest).
My only doubt about using it is regarding the performance implications it will bring, but I can't measure that properly.

---

The wrong handling of properties with the `.` somewhere that you've notice is documented here:

https://github.com/nartc/mapper/blob/29500ad0708ad81810c06ad4cf9c558e4c537aa5/packages/core/src/lib/map/specs/set.spec.ts#L22-L30

```typescript
result = set({}, 'mid.Dot', 'bar');
expect(result).toEqual({ mid: { Dot: 'bar' } });
```

it happens because `set({}, 'foo.', 'bar')` (in which the second arg is the path as `string`) assumes that the property is `foo['']` because it is a valid path.
But you're right, using array instead of string should fix that limitation.

**Comment by nartc** [comment]
Thank you both for looking into this. However, I'm pretty busy lately with work and some other Angular related open-source. If anyone of you could come up with a POC PR, I'm happy to review and discuss further.

**Comment by AliYusuf95** [comment]
@micalevisk as for performance point, I agree that proxies has impact on the performance. But they will be used only when the classes added to the map (collecting mapping properties), which is most likely will be done during the app initialization. As for the rest of the functionality the performance shouldn't be impacted.

I'll try to make PR for both features once I have a time.

**Comment by nartc** [comment]
Closed with #301
